lib/puppet/provider/user/useradd.rb

```diff
@@ -89,8 +93,39 @@ def shell=(value)
     value.is_a? Integer
   end
 
-  verify :groups, "Groups must be comma-separated" do |value|
-    value !~ /\s/
+  verify :groups, "Groups must not contain spaces" do |value|
+    prop = @resource.parameters[:groups]
+    changes = prop.should_but_isnt + prop.is_but_shouldnt
+    changes.all? do |change|
+      change !~ /\s/
+    end
+  end
+
+  def groups=(value)
+    self.class.validate(:groups, value, self)
```

Comment:
All of this is a mess. I'm bypassing the `value` param pretty much everywhere because it's the comma-joined `should` that comes from the list property.

As for the verify/validate bit, I was on the fence between doing what you suggested and trying to map my changes back to use as much of the nameservice base provider's existing functionality. I stuck with this because I wanted to make sure we kept the same error message format. I'm happy to rip out the verify/validate indirection and just throw an exception here if you don't think that's a concern.